HIP: Chart-defined plugins (#433)
helm/community: hips/hip-9999.md

---
hip: 9999
title: "H4HIP: Chart-defined plugins"
authors:
  ["Scott Rigby <scott@r6by.com>", "George Jenkins <gvjenkins@gmail.com>"]
created: "2025-04-11"
type: "feature"
status: "draft"
helm-version: "4"
---

## Abstract

This HIP builds upon [H4HIP: Wasm plugin system](./hip-0026.md) to define how charts can specify and use plugins in Helm 4. It enables chart authors to define plugins that are required for their charts to function, such as alternative render engines or download protocols for dependencies.

## Motivation

While the main plugin system HIP enables users to extend Helm's functionality, chart authors ([Application Distributors](https://github.com/helm/community/blob/main/user-profiles.md#2-application-distributor)) need a way to specify which plugins their charts require. This allows charts to use alternative render engines, download protocols, and other plugin-based functionality while ensuring users have the necessary plugins installed.

## Specification

> Support for Chart-defined plugins will be added starting in Chart API v3. `Chart.yaml` will define a top-level `plugins` key, and `Chart.lock` will capture specific plugin versions and checksums.

Reviewer: ```suggestion
Support for Chart-defined plugins will be added starting in [Chart API v3](https://github.com/helm/community/blob/main/hips/hip-0020.md). `Chart.yaml` will define a top-level `plugins` key, and `Chart.lock` will capture specific plugin versions and checksums.
```

### Chart.yaml Plugin Syntax

The `plugins` key is defined as a **list** (not a map). This is critical for `render/v1` plugins because:

- Go does not guarantee YAML map key ordering
- Sequential execution allows earlier plugins to modify a shared `SourceFiles` object in-memory
- Later plugins can then target files created or modified by earlier plugins
- This enables plugin composition where one plugin creates files and another processes them

Example `Chart.yaml` for Charts v3:

```yaml
apiVersion: v3
name: my-chart
version: 1.0.0

# Chart-defined plugins (sequential list order matters for render plugins)
plugins:
  - name: yamlscript
    type: render/v1
    repository: oci://ghcr.io/yamlscript/helm-yamlscript
    version: 0.1.0
  - name: s3
    type: getter/v1
    repository: oci://ghcr.io/hypnoglow/helm-s3
    version: 0.16.3

# Renamed from "dependencies" in v2
subcharts:
  - name: postgresql
    version: 12.x.x

>     repository: https://charts.bitnami.com/bitnami

Reviewer: I think remove the sections not relevant to plugins:

```suggestion
apiVersion: v3

# Chart-defined plugins (sequential list order matters for render plugins)
plugins:
  - name: yamlscript
    type: render/v1
    repository: oci://ghcr.io/yamlscript/helm-yamlscript
    version: 0.1.0
  - name: s3
    type: getter/v1
    repository: oci://ghcr.io/hypnoglow/helm-s3
    version: 0.16.3

```

```

> 

Reviewer 2: I feel like that detail is too implementation-specific. One possible implementation would be to leave the gotemplate rendering still as part of helm + not as a plugin, and another could be to have that as a plugin. But both possibilities aren't too important to the user.

I would word it like this:

```suggestion
When `plugin:` is unspecified, gotemplate rendering will still be available to the chart as it worked before Chart-defined plugins was implemented.
```

### Render/v1 Plugin Specification

The `render/v1` plugin type enables chart authors to use alternative rendering engines beyond gotemplate.

> > **Terminology note**: The plugin type is named `render/v1` (not `template/v1`) because not all rendering engines use templating. For example, Cue-lang uses code evaluation rather than template substitution.

Reviewer: I think we can describe this better. Something like?

```suggestion
> **Terminology note**: The plugin type is named `render/v1` (not ie. `template/v1`) to describe the process of generating Kubernetes object manifests from the Chart's source, which more generally is done by code evaluation. For example, Cue-lang uses code evaluation rather than template substitution.
```

I'm not sure where the "render" terminology comes from? Perhaps "evaluation" would be better?

    - "/templates/NOTES.txt"
```

**Example file targeting scenarios:**

- Gotemplate plugin: `["/templates/*.yaml", "/templates/*.tpl", "/templates/NOTES.txt", "/tests/*.yaml"]`
- Pkl plugin: `["/templates/*.pkl", "*.pkl.yaml"]`
- YamlScript plugin: `["/templates/*.ys", "*.ys.yaml"]`
- Kustomize plugin: `["kustomization.yaml", "overlays/*.yaml"]`
- Cue plugin: `[".cue", "*.cue.json", "*.cue.yaml"]`

This allows plugins to manage complete file dependency sets (e.g., gotemplate managing both `.tpl` partials and `.yaml` templates).

#### Plugin Invocation

All `render/v1` plugins receive the same Helm built-in objects via JSON serialization through Extism:

- `Release` - Release-specific information (name, namespace, etc.)
- `Values` - Merged values from values.yaml and --set flags
- `Chart` - Chart metadata
- `Subcharts` - Metadata about subchart dependencies
- `Files` - Non-template files accessible to the chart
- `Capabilities` - Information about the Kubernetes cluster
- `SourceFiles` - All source files in the chart (plugins can add, modify, or remove)
- `Template` - Template-specific information (name, base path) - see [Open Issue #5](#5-template-built-in-object)

> Plugins return rendered Kubernetes manifests as a map of `filename -> content`.

Reviewer 2: I was writing an example, but I'm thinking the input format may not be sufficient.

https://gist.github.com/jessesimpson36/d1d849aad060a10073205c2cc46738cd

Extism doesn't allow you to read files since it's using web assembly / doesn't have access to the filesystem. So the entire contents of the template would need to be passed in to the function. (I think. I'm no expert on Extism)

#### Sequential Execution

When multiple `render/v1` plugins are defined, they execute sequentially in list order:

1. All chart files are loaded into a shared `SourceFiles` object
2. Each plugin receives the current `SourceFiles` state

> 3. Plugin can add, modify, or remove files

Reviewer: We should describe in more detail how this interacts with "Plugins return rendered Kubernetes manifests as a map of `filename -> content`."

This unified model ensures reproducible builds while supporting version-specific plugin storage across charts.

### Helm dependency commands

The existing `helm dependency` commands are extended to handle chart-defined plugins:

- `helm dependency update` - Resolves and downloads both subcharts and plugins, updates `Chart.lock`
- `helm dependency build` - Downloads locked versions of subcharts and plugins from `Chart.lock`
- `helm dependency list` - Shows status of both subcharts and plugins

**Workflow for chart consumers:**

1. Chart author defines plugins in `Chart.yaml`
2. Run `helm dependency update` → downloads plugins to global cache, updates `Chart.lock`
3. `Chart.lock` committed to version control ensures reproducible builds
4. Other users run `helm dependency build` → installs exact locked versions
5. `helm template/install/upgrade` uses plugins from cache automatically

### Distribution

Chart-defined plugins **must be**:

1. Implemented using the **Wasm runtime** (via Extism)
2. Distributed as **OCI artifacts**

Plugins are referenced by OCI URL in `Chart.yaml` (not packaged within charts). This keeps chart sizes small while ensuring plugins are:

> - Versioned and locked via `Chart.lock` (includes SHA256 digest)

Reviewer: ```suggestion
- Versioned and locked via `Chart.lock` (lock digest corresponds to OCI manifest digest)
```

- Cached as tarballs in content-addressable storage at `$HELM_CACHE_HOME/content/`

> - Loaded directly from tarballs at runtime (no persistent extraction)

Reviewer: ```suggestion
- Loaded directly from content cache at runtime (no persistent extraction)
```

> - Shared across multiple charts using the same plugin version

Reviewer: ```suggestion
- Shared across multiple charts using the same plugin content hash
```

- Downloadable ahead of time for airgap scenarios via `helm dependency build`

**Plugin caching:** Plugin tarballs are stored in Helm's content-addressable cache (`$HELM_CACHE_HOME/content/`) keyed by SHA256 digest, matching how charts are cached. At runtime, plugins are loaded directly from the cached tarball - the `plugin.yaml` and `.wasm` files are read into memory without extracting to disk.

> **Plugin versioning:** The content cache supports multiple versions of the same plugin, allowing different charts to depend on different plugin versions without conflicts.

Reviewer: Is this trying to say, that as plugin versions are resolved to OCI manifest digests, multiple plugin versions can be supported (as they will have the same manifest digest)?

Seems a little redundant if so?

**Content-addressable caching:** Plugin tarballs are stored using their SHA256 digest, enabling:

- Deduplication: Same plugin version used by multiple charts is stored once
- Verification: Digest from `Chart.lock` validates cached content
- Consistency: Same loading pattern as charts (tarball → in-memory)
- Transience: Cache can be cleared without breaking chart workflows

**Key distinctions:**

- **Installed** plugins appear in `helm plugin list` and are available for any Helm operation
- **Cached** plugins are loaded on-demand when a chart specifies them in `Chart.yaml`
- `helm plugin list` shows only installed plugins (cached plugins are not listed)
- Installed plugins and chart-defined plugins are separate systems (no fallback between them)

**Plugin identity:**

Chart-defined plugins are identified by their **content hash** (SHA256 digest), not by name/version. This ensures deterministic builds:

- `Chart.yaml` specifies name/version for human convenience
- `helm dependency update` resolves to OCI digest, computes content hash, stores in `Chart.lock`
- At runtime, Helm loads the plugin by content hash from `Chart.lock`

> - If the hash is not in the cache, the plugin is downloaded

Reviewer: This should be clarified / worded differently I think.

(note for myself, I will propose some updates)
